spmap/spmap.go

```diff
@@ -0,0 +1,138 @@
+// spmap is a map of [peerid->spid]
+// when the mapping is not known locally, hey-fil is used to find it.
+package spmap
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"sync"
+
+	"github.com/ipfs/go-log/v2"
+	"github.com/libp2p/go-libp2p/core/peer"
+)
+
+var logger = log.Logger("lassie/spmap")
+
+type Option func(spConfig)
+
+func NewSPMap(opts ...Option) *SPMap {
+	cf := spConfig{
+		heyFilEndpoint: "https://heyfil.prod.cid.contact",
+		client:         http.DefaultClient,
+	}
+	for _, o := range opts {
+		o(cf)
+	}
+	sm := SPMap{
+		cfg:   cf,
+		cache: map[string][]string{},
+		c:     make(chan work, 10),
+	}
+	go sm.run()
+	return &sm
+}
+
+type spConfig struct {
+	heyFilEndpoint string
+	client         *http.Client
+}
+
+func WithHeyFil(endpoint string) Option {
+	return func(sc spConfig) {
+		sc.heyFilEndpoint = endpoint
+	}
+}
+
+func WithClient(c *http.Client) Option {
+	return func(sc spConfig) {
+		sc.client = c
+	}
+}
+
+type SPMap struct {
+	cfg spConfig
+
+	cache map[string][]string
+	lk    sync.RWMutex
+
+	c chan work
+}
+
+type work struct {
+	query    peer.ID
+	response chan string
+}
+
+func (s *SPMap) get(id string) ([]string, bool) {
+	s.lk.RLock()
+	defer s.lk.RUnlock()
+	v, ok := s.cache[id]
+	return v, ok
+}
+
+func (s *SPMap) set(id string, val []string) {
+	s.lk.Lock()
```

Comment:
Check if slice is not empty before storing?

Reply:
added comment